### DISCUSSION TITLE
`startTransition` never commit to transitioned state branch if promise not resolved for 5000MS

### DISCUSSION THREAD

**Discussion by ellemedit** [vercel/next.js#53215] [Help]
### Summary

[Reproduce codesandbox](https://codesandbox.io/p/sandbox/blissful-sky-3pm4l7?file=%2Fapp%2FClientButton.tsx%3A21%2C40)

<img width="1348" alt="image" src="https://github.com/vercel/next.js/assets/2682865/88aaa158-4c36-451f-9226-c89e94529e06">

In above example, 0~4900MS delay startTransition works well. But if you click `Wait in 5000MS` you can see, never reaching completion of pending status.

I expected no time limitation in startTransition async action. Is this a spec?

### Example

https://codesandbox.io/p/sandbox/blissful-sky-3pm4l7?file=%2Fapp%2FClientButton.tsx%3A21%2C40

**Comment by ellemedit** [comment]
I can see reproduction in latest next.js

**Comment by icyJoseph** [comment]
While it is strange that there's a number breakpoint, I'd say that this is a bit of a misusage of transitions.

- https://react.dev/reference/react/useTransition#react-doesnt-treat-my-state-update-as-a-transition

We see from the React docs that the transition callback has to be synchronous.

For example:

```ts
startTransition(() => {
  // ❌ Setting state *after* startTransition call
  setTimeout(() => {
    setPage('/about');
  }, 1000);
});
```

Which should be:

```ts
setTimeout(() => {
  startTransition(() => {
    // ✅ Setting state *during* startTransition call
    setPage('/about');
  });
}, 1000);
```

**Reply by ellemedit** [reply]
I agree this is mistake if example is in current stable React release.

But I believe canary React release doesn't, I think this is minimal reproduction of server action which is coming in next.js.

**Reply by ellemedit** [reply]
React canary already supports async action

**Comment by ellemedit** [comment]
This is hurting UX. If a user submitted form with `startTransition`, it can lead infinite loading state. It can't be controlled as API/IO task does. React actually doesn't commit prerendered branch from transitioning state.
